### DISCUSSION TITLE
Better ways of mapping records to structs?

### DISCUSSION THREAD

**Discussion by HiImJulien** [pocketbase/pocketbase#5192] [Q&A] [answered]
Hey there!

I am using Pocketbase as a Go Framework. Truth be told, I am new to go (although not new to programming).

In a custom route I am fetching the user's record as described in the documentation:

```go
userRecord,  _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
```

Now I have a custom struct, which I populate with a custom function:

```go
type UserInfo struct {
	FirstName    string `liquid:"firstName"`
	LastName     string `liquid:"lastName"`
	BusinessName string `liquid:"businessName"`
	Street       string `liquid:"street"`
	ZipCode      string `liquid:"zipCode"`
	City         string `liquid:"city"`
	Country      string `liquid:"country"`
	BankName     string `liquid:"bank"`
	BankIBAN     string `liquid:"iban"`
	BankID       string `liquid:"bic"`
	TaxID        string `liquid:"taxId"`
}

func getUserInfoFromRecord(record *models.Record) UserInfo {
	return UserInfo{
		FirstName:    record.GetString("firstName"),
		LastName:     record.GetString("lastName"),
		BusinessName: record.GetString("businessName"),
		Street:       record.GetString("street"),
		ZipCode:      record.GetString("zipCode"),
		City:         record.GetString("city"),
		Country:      record.GetString("country"),
		BankName:     record.GetString("bankName"),
		BankIBAN:     record.GetString("bankIBAN"),
		BankID:       record.GetString("bankBIC"),
		TaxID:        record.GetString("taxId"),
	}
}
```

Is there a better to populate the struct from the record than manually calling `GetXXX` for each property?

I tried to search previous issues and discussion but failed to find anything. Perhaps I am lacking the right go lingo.

**Comment by ganigeorgiev** [comment, marked as answer]
At the moment there is no automated way that will produce a typed struct from a `Record`, if that's what you are looking for.

An alternative of manually assigning each prop (but not very performant one) could be to call `json.Marshal(record)` and then `json.Unmarshal(marshaledRecord, &userInfo)` for your custom struct (_you may also have to add the `json` struct tags to each of your props_).

With the [next release](https://github.com/pocketbase/pocketbase/discussions/4355) there will be an option to autogenerate "Proxy"/"Projection" Go structs from the collections based on the suggestion from https://github.com/pocketbase/pocketbase/discussions/2697#discussioncomment-6154067 but even then I'm not sure that this will satisfy your specific use case due to the used `liquid` struct tag.

**Reply by ganigeorgiev** [reply]
If you use the above struct just so that you can pass it in a [osteele/liquid](https://github.com/osteele/liquid) template, then another alternative could be to directly export the record as map using [`record.PublicExport()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.16/models#Record.PublicExport).

**Reply by HiImJulien** [reply]
Thanks for the quick reply!

Given that the other solution is not very performant, I'll stick with what I have at the moment.

It just felt like I am doing it wrong, which was I considered likely, as I barely know golang.
